pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,33 +288,74 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+type candidateList struct {
+	idx   int32
+	items []Candidate
+}
+
+func newCandidateList(size int32) *candidateList {
+	return &candidateList{idx: -1, items: make([]Candidate, size)}
+}
+
+// add adds a new candidate to the internal array atomically.
+func (cl *candidateList) add(c *candidate) {
+	if idx := atomic.AddInt32(&cl.idx, 1); idx < int32(len(cl.items)) {
+		cl.items[idx] = c
+	}
+}
+
+// size returns the number of candidate stored atomically.
+func (cl *candidateList) size() int32 {
+	n := atomic.LoadInt32(&cl.idx) + 1
+	if n >= int32(len(cl.items)) {
+		n = int32(len(cl.items))
+	}
+	return n
+}
+
+// get returns the internal candidate array. This function is NOT atomic and
+// assumes that the caller has acquired mutual exclusion on the list.
+func (cl *candidateList) get() []Candidate {
+	return cl.items[:cl.size()]
```

Comment:
returning a subslice here means that callers can modify members of `cl.items` beyond `cl.size()` which will be overwritten if cl.add is called again (see https://play.golang.org/p/Uw_cq4NK4Tz)

Reply:
IMO this isn't a concern. `get()` is guaranteed to be called after all `add()` operations, but perhaps it's worth adding this as a note to the function documentation -- done.

Reply:
making the other methods atomically safe implies multiple goroutines are interacting with this object... are all parallel operations guaranteed to be stopped before this is called or can a timeout return control to the get() caller while some background operations are still running?

Reply:
>are all parallel operations guaranteed to be stopped before this is called

Yes, parallelize.Until will guarantee that all goroutines interacting with this candidateList will have completed before this called.